finality-aleph/src/session_map.rs

```diff
@@ -0,0 +1,528 @@
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    client: Arc<C>,
+    _phantom: PhantomData<(B, BE)>,
+}
+
+impl<C, B, BE> AuthorityProviderImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<C, B, BE> AuthorityProvider<NumberFor<B>> for AuthorityProviderImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    fn authorities(&self, num: NumberFor<B>) -> Option<Vec<AuthorityId>> {
+        self.client
+            .runtime_api()
+            .authorities(&BlockId::Number(num))
+            .ok()
+    }
+
+    fn next_authorities(&self, num: NumberFor<B>) -> Option<Vec<AuthorityId>> {
+        self.client
+            .runtime_api()
+            .next_session_authorities(&BlockId::Number(num))
+            .map(|r| r.ok())
+            .ok()
+            .flatten()
+    }
+}
+
+pub trait FinalityNotificator<B, N> {
+    fn notification_stream(&mut self) -> TracingUnboundedReceiver<B>;
+    fn last_finalized(&self) -> N;
+}
+
+/// Default implementation of finality notificator trait.
+pub struct FinalityNotificatorImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    client: Arc<C>,
+    _phantom: PhantomData<(B, BE)>,
+}
+
+impl<C, B, BE> FinalityNotificatorImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    pub fn new(client: Arc<C>) -> Self {
+        Self {
+            client,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<C, B, BE> FinalityNotificator<FinalityNotification<B>, NumberFor<B>>
+    for FinalityNotificatorImpl<C, B, BE>
+where
+    C: ClientForAleph<B, BE> + Send + Sync + 'static,
+    C::Api: aleph_primitives::AlephSessionApi<B>,
+    B: Block,
+    BE: Backend<B> + 'static,
+{
+    fn notification_stream(&mut self) -> TracingUnboundedReceiver<FinalityNotification<B>> {
+        self.client.finality_notification_stream()
+    }
+
+    fn last_finalized(&self) -> NumberFor<B> {
+        self.client.info().finalized_number
+    }
+}
+
+#[derive(Clone)]
+/// Wrapper around Mapping from sessionId to Vec of AuthorityIds allowing mutation
+/// and hiding locking details
+struct SharedSessionMap(Arc<Mutex<(SessionMap, SessionSubscribers)>>);
+
+#[derive(Clone)]
+/// Wrapper around Mapping from sessionId to Vec of AuthorityIds allowing only reads
+pub struct ReadOnlySessionMap {
+    inner: Arc<Mutex<(SessionMap, SessionSubscribers)>>,
```

Comment:
maybe a good candidate for a use of RwLock, if you see some read concurrency ever happening?

Reply:
done